lib/proxy/conn.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+)
+
+const (
+	// bufferSize is the connection buffer size.
+	bufferSize = 4 * 1024
```

Comment:
Any specific reason to choose 4KiB?

Reply:
Nope kind of an arbitrary choice. But now that I think about it I could rewrite some of this to get rid of the buffer/net.Pipe and have streamConn.Read read from the grpc stream directly.

Reply:
Seems like doing this refactoring would simplify things a bit. @dboslee did you look into this, is it something you plan on doing?

Reply:
Updated this here https://github.com/gravitational/teleport/pull/10223/commits/1d3e96f59d3896368fb022ae29ca97971e585254